File: Python/algorithmic_heights/18_hs.py
# Heap Sort
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def get_index_of_parent_children_maximum(array, parent_ix, size):
    max_ix = parent_ix
    left_child_ix = 2*parent_ix + 1
    if left_child_ix < size and array[left_child_ix] > array[max_ix]:
        max_ix = left_child_ix

    right_child_ix = left_child_ix + 1
    if right_child_ix < size and array[right_child_ix] > array[max_ix]:
        max_ix = right_child_ix

    return max_ix


def sift_down(array, parent_ix, size):
    while (max_ix := get_index_of_parent_children_maximum(array, parent_ix, size)) != parent_ix:
        array[max_ix], array[parent_ix] = array[parent_ix], array[max_ix]
        parent_ix = max_ix


def heapify(array, size):
    parent_ix = size // 2 - 1
    while parent_ix >= 0:
        sift_down(array, parent_ix, size)
        parent_ix -= 1


def heapsort(array, size):
    heapify(array, size)
    n = size
    while n > 1:
        array[0], array[n-1] = array[n-1], array[0]
        sift_down(array, 0, n-1)
        n -= 1


def main():
    reader = sys.stdin
    n = int(next(reader))
    array = convert_to_intlist(next(reader))
    heapsort(array, n)
    print(' '.join(map(str, array)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/18_hs.hs
import Control.Monad (mapM_, when)
import Control.Monad.ST (runST)
import Data.Vector (Vector, freeze, fromList, thaw, toList)
import qualified Data.Vector.Mutable as MV

convertToIntVector :: String -> Vector Int
convertToIntVector = fromList . map read . words


heapsort :: Ord a => Vector a -> Int -> Vector a
heapsort vector n =
    let getIndexOfParentChildrenMaximum array parentIx n = do
            let leftChildIx = 2 * parentIx + 1
            let rightChildIx = leftChildIx + 1
            maxIx <- if leftChildIx < n then do
                        x <- MV.read array leftChildIx
                        y <- MV.read array parentIx
                        return $ if x > y then leftChildIx else parentIx
                    else return parentIx
            if rightChildIx < n then do
                x <- MV.read array rightChildIx
                y <- MV.read array maxIx
                return $ if x > y then rightChildIx else maxIx
            else return maxIx

        siftDown array parentIx n =
            let go currentParentIx = do
                    maxIx <- getIndexOfParentChildrenMaximum array currentParentIx n
                    when (maxIx /= currentParentIx) $ do
                        MV.swap array maxIx currentParentIx
                        go maxIx
            in go parentIx

        createHeap array parentIx
            | parentIx >= 0 = do
                siftDown array parentIx n
                createHeap array (parentIx - 1)
            | otherwise = return ()

        moveElemToItsPlace array k = do
            MV.swap array 0 (k - 1)
            siftDown array 0 (k - 1)

    in runST $ do
        array <- thaw vector
        createHeap array (n `div` 2 - 1)
        mapM_ (moveElemToItsPlace array) [n, (n-1)..2]
        freeze array


main :: IO ()
main = do
    n <- readLn
    vector <- convertToIntVector <$> getLine
    let result = heapsort vector n
    putStrLn $ unwords $ map show $ toList result
